src/log.rs

```diff
@@ -81,13 +84,327 @@ type CompatFilterStream<T, R> = Compat01As03<FilterStream<T, R>>;
+    /// builder's parameters.
+    pub fn past_logs_pages(self) -> impl Stream<Item = Result<Vec<Log>, ExecutionError>> {
+        stream::try_unfold(PastLogsStream::Init(self), PastLogsStream::next)
+            .try_filter(|logs| future::ready(!logs.is_empty()))
+    }
+}
+
+/// Internal unfold context for creating a `past_logs` `Stream`.
+enum PastLogsStream<T: Transport> {
+    Init(LogFilterBuilder<T>),
+    Done,
+    Paging(PastLogsPager<T>),
+    Querying(Web3<T>, Filter),
+}
+
+impl<T: Transport> PastLogsStream<T> {
+    async fn next(mut self) -> Result<Option<(Vec<Log>, Self)>, ExecutionError> {
+        loop {
+            let (logs, next) = match self {
+                PastLogsStream::Init(builder) => {
+                    self = PastLogsStream::init(builder).await?;
+                    continue;
+                }
+                PastLogsStream::Done => return Ok(None),
+                PastLogsStream::Paging(mut pager) => {
+                    let logs = match pager.next_page().await? {
+                        Some(logs) => logs,
+                        None => return Ok(None),
+                    };
+                    (logs, PastLogsStream::Paging(pager))
+                }
+                PastLogsStream::Querying(web3, filter) => {
+                    let logs = web3.eth().logs(filter.clone()).compat().await?;
+                    (logs, PastLogsStream::Done)
+                }
+            };
+            return Ok(Some((logs, next)));
+        }
+    }
+
+    async fn init(builder: LogFilterBuilder<T>) -> Result<Self, ExecutionError> {
+        let from_block = builder.from_block.unwrap_or(BlockNumber::Latest);
+        let to_block = builder.to_block.unwrap_or(BlockNumber::Latest);
+
+        let web3 = builder.web3.clone();
+        let block_page_size = builder
+            .block_page_size
+            .map(|size| size.get())
+            .unwrap_or(DEFAULT_BLOCK_PAGE_SIZE);
+        let filter = builder.into_filter();
+
+        let start_block = match from_block {
+            BlockNumber::Earliest => Some(0),
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => None,
+        };
+        let end_block = match to_block {
+            BlockNumber::Earliest => None,
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => {
+                let latest_block = web3.eth().block_number().compat().await?;
+                Some(latest_block.as_u64())
+            }
+        };
+
+        let next = match (start_block, end_block) {
+            (Some(page_block), Some(end_block)) => PastLogsStream::Paging(PastLogsPager {
+                web3,
+                to_block,
+                block_page_size,
+                filter,
+                page_block,
+                end_block,
+            }),
+            _ => PastLogsStream::Querying(web3, filter.build()),
```

Comment:
What is the point of Querying? Isn't this the same as doing Paging once if we would set `start_block` to the same value as `end_block` above?

Reply:
This branch is required for when we can't determine the initial `page_block` or `end_block`, and want to do a single query to the node with the provided `to_block` and `from_block`.

This mostly stems from the fact that you can specify some things that don't really make sense range wise, for example: `filter.from_block(BlockNumber::Pending).to_block(BlockNumber::Earliest)`. What does this query even mean? Instead of coming up with errors, I just make a query and differ the decision on how to handle this case to the node.

Reply:
I see. Could also argue that we shouldn't allow this and get rid of more code but I don't mind either way.

Reply:
I'll keep it as-is so it works like the `query_past_logs_paginated` used to. It also short-circuits the case where you do:
```
LogFilterBuilder::new(web3).past_log_pages();
// equivalent to
LogFilterBuilder::new(web3).from_block(Latest).to_block(Latest).past_log_pages();
```

Which should work and not require fetching the `end_block`.